clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```
@@ -1069,6 +1069,13 @@ private HeartbeatResponseHandler(final Generation generation) {
         public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> future) {
             sensors.heartbeatSensor.record(response.requestLatencyMs());
             Errors error = heartbeatResponse.error();
+
+            if (state != MemberState.STABLE) {
```

We should still handle fatal exception IMHO, such as FencedInstanceIdException

---

I can see it either way. It seems like this PR is about sending the heartbeats _optimistically_ during rebalance, so there doesn't seem to really be any harm in ignoring the response for now. If we ignore the errors, then everything should still work, as the JoinGroup or SyncGroup response will tell us that we've been fenced next time we poll.

It seems like the advantage of handling the error here is that we can potentially rejoin just a tiny bit sooner by not having to wait for the JoinGroup or SyncGroup response. But it's not clear to me that it's actually ok not to handle those responses, so then we would also need to make sure the response handling logic can detect that the response has already been invalidated if we've sent a new JoinGroup request in the mean time.

This definitely has the potential to decrease the MTTR, but I'm wondering if we should take on the complexity right now, or consider it as a follow-on optimization.

---

I agree with @vvcephei , we should aim to keep this fix as simple as possible. The reason for this PR is that members are getting kicked from the group unnecessarily, not that members who are actually supposed to be fenced don't rejoin fast enough.
If we fix KAFKA-10122 and people start complaining that members fenced during a rebalance take too long to find out, then we can revisit this optimization. But if KAFKA-10122 is fixed then it seems like the occurrence of members dropping out during a rebalance will go way down anyway, so we'd be solving a vanishingly small problem

---

My thoughts were that, when we are in rebalancing then the purpose of heartbeat is only to keep the consumer alive at the broker side, not to take any instructions. But I think it should be handled case-by-case, I will try to refactor this piece a bit as well.